File: components/cell-mesh/src/cell_mnca.c
/**
 * cell_mnca.c — MNCA tile compute + incentive quorum.
 * Pure C, no IDF dependency — host-testable.
 */

#include "cell_mnca.h"
#include <string.h>
#include <stdint.h>
#include <stdbool.h>

// ── Reference rule ─────────────────────────────────────────────────────────────
// Conway-like: birth at 3, survive at 2-3, integer state (grow +64 / decay -64).
// rule_id[4] = { 'M','N','C','A' } (ASCII, identifies this rule family).
const cm_mnca_rule_t CM_MNCA_DEFAULT_RULE = {
    .alive_threshold = 128,
    .inner_radius    = 1,
    .birth_lo        = 3,
    .birth_hi        = 3,
    .survive_lo      = 2,
    .survive_hi      = 3,
    .grow_step       = 64,
    .decay_step      = 64,
    .rule_id         = { 'M', 'N', 'C', 'A' },
};

// ── Tile helpers ───────────────────────────────────────────────────────────────

void cm_mnca_tile_init_random(cm_mnca_tile_t *t, uint16_t x, uint16_t y, uint32_t seed) {
    if (!t) return;
    t->x          = x;
    t->y          = y;
    t->generation = 0;
    // LCG: simple deterministic "random" seed-based init.
    uint32_t s = seed ^ 0xdeadbeef;
    for (size_t i = 0; i < CM_MNCA_TILE_CELLS; i++) {
        s = s * 1664525u + 1013904223u;
        t->state[i] = (uint8_t)(s >> 16);
    }
}

static int count_alive(const uint8_t *cells, int w, int h, int cx, int cy,
                       int radius, uint8_t thresh) {
    int n = 0;
    for (int dy = -radius; dy <= radius; dy++) {
        for (int dx = -radius; dx <= radius; dx++) {
            if (dx == 0 && dy == 0) continue;
            int nx = cx + dx, ny = cy + dy;
            if (nx < 0 || nx >= w || ny < 0 || ny >= h) continue;
            if (cells[ny * w + nx] >= thresh) n++;
        }
    }
    return n;
}

static uint8_t clamp_u8(int v) {
    return v < 0 ? 0 : v > 255 ? 255 : (uint8_t)v;
}

int cm_mnca_step(const cm_mnca_tile_t *cur, cm_mnca_tile_t *next,
                  const cm_mnca_rule_t *rule) {
    if (!cur || !next || !rule) return -1;
    *next = *cur;
    next->generation = cur->generation + 1;
    const int W = CM_MNCA_TILE_W, H = CM_MNCA_TILE_H;
    for (int y = 0; y < H; y++) {
        for (int x = 0; x < W; x++) {
            uint8_t self   = cur->state[y * W + x];
            int     alive  = count_alive(cur->state, W, H, x, y,
                                         rule->inner_radius, rule->alive_threshold);
            bool    is_alive = (self >= rule->alive_threshold);
            int     delta;
            if (is_alive) {
                delta = (alive >= rule->survive_lo && alive <= rule->survive_hi)
                       ? rule->grow_step : -(int)rule->decay_step;
            } else {
                delta = (alive >= rule->birth_lo && alive <= rule->birth_hi)
                       ? rule->grow_step : -(int)rule->decay_step;
            }
            next->state[y * W + x] = clamp_u8((int)self + delta);
        }
    }
    return 0;
}

// ── Payload encode / decode ────────────────────────────────────────────────────

static void write_u16_le(uint8_t *p, uint16_t v) { p[0] = v & 0xff; p[1] = v >> 8; }
static void write_u32_le(uint8_t *p, uint32_t v) {
    p[0] = v & 0xff; p[1] = (v >> 8) & 0xff;
    p[2] = (v >> 16) & 0xff; p[3] = (v >> 24) & 0xff;
}
static uint16_t read_u16_le(const uint8_t *p) { return (uint16_t)p[0] | ((uint16_t)p[1] << 8); }
static uint32_t read_u32_le(const uint8_t *p) {
    return (uint32_t)p[0] | ((uint32_t)p[1] << 8) | ((uint32_t)p[2] << 16) | ((uint32_t)p[3] << 24);
}

size_t cm_mnca_tile_encode(const cm_mnca_tile_t *t, const cm_mnca_rule_t *rule,
                            uint8_t out[CM_PAYLOAD_SIZE]) {
    if (!t || !out) return 0;
    memset(out, 0, CM_PAYLOAD_SIZE);
    write_u16_le(out + CM_MNCA_TILE_V0_OFF_X,       t->x);
    write_u16_le(out + CM_MNCA_TILE_V0_OFF_Y,       t->y);
    write_u32_le(out + CM_MNCA_TILE_V0_OFF_GEN,     t->generation);
    if (rule) memcpy(out + CM_MNCA_TILE_V0_OFF_RULE_ID, rule->rule_id, 4);
    write_u32_le(out + CM_MNCA_TILE_V0_OFF_STATE_LEN, (uint32_t)CM_MNCA_TILE_CELLS);
    memcpy(out + CM_MNCA_TILE_V0_OFF_STATE, t->state, CM_MNCA_TILE_CELLS);
    return CM_MNCA_TILE_V0_HDR_BYTES + CM_MNCA_TILE_CELLS;
}

int cm_mnca_tile_decode(const uint8_t *payload, size_t payload_len, cm_mnca_tile_t *out) {
    if (!payload || !out) return -1;
    if (payload_len < CM_MNCA_TILE_V0_HDR_BYTES) return -1;
    uint32_t state_len = read_u32_le(payload + CM_MNCA_TILE_V0_OFF_STATE_LEN);
    if ((size_t)(CM_MNCA_TILE_V0_HDR_BYTES + state_len) > payload_len) return -1;
    if (state_len != CM_MNCA_TILE_CELLS) return -1;  // expect 8×8 in this impl
    out->x          = read_u16_le(payload + CM_MNCA_TILE_V0_OFF_X);
    out->y          = read_u16_le(payload + CM_MNCA_TILE_V0_OFF_Y);
    out->generation = read_u32_le(payload + CM_MNCA_TILE_V0_OFF_GEN);
    memcpy(out->state, payload + CM_MNCA_TILE_V0_OFF_STATE, CM_MNCA_TILE_CELLS);
    return 0;
}

// ── SHA-256 (pure C — no mbedTLS, for host-testability) ──────────────────────
// This is a compact self-contained SHA-256 implementation used ONLY by the
// quorum hash path in host tests.  The firmware can replace with mbedtls_sha256.

static const uint32_t K256[64] = {
    0x428a2f98u,0x71374491u,0xb5c0fbcfu,0xe9b5dba5u,
    0x3956c25bu,0x59f111f1u,0x923f82a4u,0xab1c5ed5u,
    0xd807aa98u,0x12835b01u,0x243185beu,0x550c7dc3u,
    0x72be5d74u,0x80deb1feu,0x9bdc06a7u,0xc19bf174u,
    0xe49b69c1u,0xefbe4786u,0x0fc19dc6u,0x240ca1ccu,
    0x2de92c6fu,0x4a7484aau,0x5cb0a9dcu,0x76f988dau,
    0x983e5152u,0xa831c66du,0xb00327c8u,0xbf597fc7u,
    0xc6e00bf3u,0xd5a79147u,0x06ca6351u,0x14292967u,
    0x27b70a85u,0x2e1b2138u,0x4d2c6dfcu,0x53380d13u,
    0x650a7354u,0x766a0abbu,0x81c2c92eu,0x92722c85u,
    0xa2bfe8a1u,0xa81a664bu,0xc24b8b70u,0xc76c51a3u,
    0xd192e819u,0xd6990624u,0xf40e3585u,0x106aa070u,
    0x19a4c116u,0x1e376c08u,0x2748774cu,0x34b0bcb5u,
    0x391c0cb3u,0x4ed8aa4au,0x5b9cca4fu,0x682e6ff3u,
    0x748f82eeu,0x78a5636fu,0x84c87814u,0x8cc70208u,
    0x90befffau,0xa4506cebu,0xbef9a3f7u,0xc67178f2u,
};

#define ROTR(x,n) (((x) >> (n)) | ((x) << (32-(n))))
#define S0(x) (ROTR(x,2)  ^ ROTR(x,13) ^ ROTR(x,22))
#define S1(x) (ROTR(x,6)  ^ ROTR(x,11) ^ ROTR(x,25))
#define G0(x) (ROTR(x,7)  ^ ROTR(x,18) ^ ((x) >> 3))
#define G1(x) (ROTR(x,17) ^ ROTR(x,19) ^ ((x) >> 10))
#define CH(x,y,z) (((x)&(y))^(~(x)&(z)))
#define MAJ(x,y,z) (((x)&(y))^((x)&(z))^((y)&(z)))

static void sha256_block(uint32_t H[8], const uint8_t block[64]) {
    uint32_t W[64], a,b,c,d,e,f,g,h,T1,T2;
    for (int i = 0; i < 16; i++) {
        W[i] = ((uint32_t)block[4*i+0]<<24)|((uint32_t)block[4*i+1]<<16)
              |((uint32_t)block[4*i+2]<<8)|(uint32_t)block[4*i+3];
    }
    for (int i = 16; i < 64; i++) W[i] = G1(W[i-2])+W[i-7]+G0(W[i-15])+W[i-16];
    a=H[0];b=H[1];c=H[2];d=H[3];e=H[4];f=H[5];g=H[6];h=H[7];
    for (int i = 0; i < 64; i++) {
        T1 = h + S1(e) + CH(e,f,g) + K256[i] + W[i];
        T2 = S0(a) + MAJ(a,b,c);
        h=g;g=f;f=e;e=d+T1;d=c;c=b;b=a;a=T1+T2;
    }
    H[0]+=a;H[1]+=b;H[2]+=c;H[3]+=d;H[4]+=e;H[5]+=f;H[6]+=g;H[7]+=h;
}

void cm_mnca_tile_hash(const cm_mnca_tile_t *t, uint8_t out_hash[32]) {
    if (!t || !out_hash) return;
    const uint8_t *data = t->state;
    uint32_t len = CM_MNCA_TILE_CELLS;
    uint32_t H[8] = {
        0x6a09e667u,0xbb67ae85u,0x3c6ef372u,0xa54ff53au,
        0x510e527fu,0x9b05688cu,0x1f83d9abu,0x5be0cd19u,
    };
    uint8_t block[64];
    uint32_t pos = 0;
    while (pos < len) {
        uint32_t copy = len - pos;
        if (copy > 64) copy = 64;
        if (copy < 64) {
            memcpy(block, data + pos, copy);
            block[copy] = 0x80;
            memset(block + copy + 1, 0, 64 - copy - 1);
            if (copy < 56) {
                uint64_t bitlen = (uint64_t)len * 8;
                for (int i = 0; i < 8; i++) block[63-i] = (uint8_t)(bitlen >> (8*i));
                sha256_block(H, block);
            } else {
                sha256_block(H, block);
                memset(block, 0, 56);
                uint64_t bitlen = (uint64_t)len * 8;
                for (int i = 0; i < 8; i++) block[63-i] = (uint8_t)(bitlen >> (8*i));
                sha256_block(H, block);
            }
            break;
        }
        memcpy(block, data + pos, 64);
        sha256_block(H, block);
        pos += 64;
    }
    if (len == 0) {
        memset(block, 0, 64); block[0] = 0x80;
        uint64_t bitlen = 0;
        for (int i = 0; i < 8; i++) block[63-i] = 0;
        sha256_block(H, block);
    }
    for (int i = 0; i < 8; i++) {
        out_hash[4*i+0] = (uint8_t)(H[i]>>24);
        out_hash[4*i+1] = (uint8_t)(H[i]>>16);
        out_hash[4*i+2] = (uint8_t)(H[i]>>8);
        out_hash[4*i+3] = (uint8_t)(H[i]);
    }
}

// ── Quorum table ───────────────────────────────────────────────────────────────

void cm_mnca_quorum_init(cm_mnca_quorum_t *q) {
    if (!q) return;
    memset(q, 0, sizeof(*q));
}

cm_mnca_quorum_rc_t cm_mnca_quorum_update(cm_mnca_quorum_t *q,
                                           uint16_t x, uint16_t y, uint32_t generation,
                                           const uint8_t tile_hash[32],
                                           const uint8_t sender_mac[6],
                                           uint64_t now_ms) {
    if (!q || !tile_hash || !sender_mac) return CM_MNCA_QUORUM_PENDING;

    // Evict stale slots.
    for (int i = 0; i < (int)CM_MNCA_QUORUM_SLOTS; i++) {
        cm_mnca_quorum_slot_t *s = &q->slots[i];
        if (s->valid && (now_ms - s->first_seen_ms) > CM_MNCA_QUORUM_TTL_MS)
            s->valid = false;
    }

    // Find an existing slot for (x, y, generation) or claim a free one.
    cm_mnca_quorum_slot_t *slot = NULL;
    int free_idx = -1;
    for (int i = 0; i < (int)CM_MNCA_QUORUM_SLOTS; i++) {
        cm_mnca_quorum_slot_t *s = &q->slots[i];
        if (!s->valid) { if (free_idx < 0) free_idx = i; continue; }
        if (s->x == x && s->y == y && s->generation == generation) { slot = s; break; }
    }
    if (!slot) {
        if (free_idx < 0) return CM_MNCA_QUORUM_PENDING;  // no room
        slot = &q->slots[free_idx];
        slot->valid        = true;
        slot->x            = x;
        slot->y            = y;
        slot->generation   = generation;
        slot->seen_count   = 0;
        slot->first_seen_ms = now_ms;
    }

    // Ignore duplicate sender for same (x, y, gen) — replay protection.
    for (uint8_t j = 0; j < slot->seen_count; j++) {
        if (memcmp(slot->sender_mac[j], sender_mac, 6) == 0) return CM_MNCA_QUORUM_PENDING;
    }

    // Record this observation.
    uint8_t idx = slot->seen_count;
    if (idx >= CM_MNCA_QUORUM_K + 1) return CM_MNCA_QUORUM_PENDING;  // overflow guard
    memcpy(slot->tile_hash[idx], tile_hash, 32);
    memcpy(slot->sender_mac[idx], sender_mac, 6);
    slot->seen_count++;

    // Check if K entries share the same hash.
    for (uint8_t a = 0; a < slot->seen_count; a++) {
        uint8_t match = 1;
        for (uint8_t b = a + 1; b < slot->seen_count; b++) {
            if (memcmp(slot->tile_hash[a], slot->tile_hash[b], 32) == 0) {
                match++;
            }
        }
        if (match >= CM_MNCA_QUORUM_K) {
            slot->valid = false;  // fire-once — invalidate
            return CM_MNCA_QUORUM_HIT;
        }
    }
    return CM_MNCA_QUORUM_PENDING;
}

File: components/cell-mesh-zig/src/cell_mnca.zig
const std = @import("std");
const Sha256 = std.crypto.hash.sha2.Sha256;
const wire = @import("cell_wire.zig");

pub const tile_header_bytes: usize = 16;
pub const max_state_bytes: usize = wire.payload_size - tile_header_bytes;
pub const tile_w: usize = 8;
pub const tile_h: usize = 8;
pub const tile_cells: usize = tile_w * tile_h;
pub const quorum_slots: usize = 4;
pub const quorum_k: usize = 2;
pub const quorum_ttl_ms: u64 = 60000;

pub const MncaRule = extern struct {
    alive_threshold: u8,
    inner_radius: u8,
    birth_lo: u8,
    birth_hi: u8,
    survive_lo: u8,
    survive_hi: u8,
    grow_step: u8,
    decay_step: u8,
    rule_id: [4]u8,
};

pub export const CM_MNCA_DEFAULT_RULE: MncaRule = .{
    .alive_threshold = 128,
    .inner_radius = 1,
    .birth_lo = 3,
    .birth_hi = 3,
    .survive_lo = 2,
    .survive_hi = 3,
    .grow_step = 64,
    .decay_step = 64,
    .rule_id = .{ 'M', 'N', 'C', 'A' },
};

pub const Tile = extern struct {
    x: u16,
    y: u16,
    generation: u32,
    state: [tile_cells]u8,
};

pub const QuorumSlot = extern struct {
    valid: bool,
    x: u16,
    y: u16,
    generation: u32,
    seen_count: u8,
    tile_hash: [quorum_k + 1][32]u8,
    sender_mac: [quorum_k + 1][6]u8,
    first_seen_ms: u64,
};

pub const Quorum = extern struct {
    slots: [quorum_slots]QuorumSlot,
};

pub export fn cm_mnca_tile_init_random(maybe_t: ?*Tile, x: u16, y: u16, seed: u32) callconv(.c) void {
    const t = maybe_t orelse return;
    t.x = x;
    t.y = y;
    t.generation = 0;
    var s = seed ^ 0xDEADBEEF;
    for (&t.state) |*cell| {
        s = s *% 1664525 +% 1013904223;
        cell.* = @intCast((s >> 16) & 0xff);
    }
}

fn countAlive(cells: *const [tile_cells]u8, cx: i32, cy: i32, radius: u8, thresh: u8) i32 {
    var n: i32 = 0;
    var dy: i32 = -@as(i32, radius);
    while (dy <= radius) : (dy += 1) {
        var dx: i32 = -@as(i32, radius);
        while (dx <= radius) : (dx += 1) {
            if (dx == 0 and dy == 0) continue;
            const nx = cx + dx;
            const ny = cy + dy;
            if (nx < 0 or nx >= @as(i32, @intCast(tile_w)) or ny < 0 or ny >= @as(i32, @intCast(tile_h))) continue;
            const idx: usize = @intCast(ny * @as(i32, @intCast(tile_w)) + nx);
            if (cells[idx] >= thresh) n += 1;
        }
    }
    return n;
}

fn clampU8(v: i32) u8 {
    if (v < 0) return 0;
    if (v > 255) return 255;
    return @intCast(v);
}

pub export fn cm_mnca_step(maybe_cur: ?*const Tile, maybe_next: ?*Tile, maybe_rule: ?*const MncaRule) callconv(.c) c_int {
    const cur = maybe_cur orelse return -1;
    const next = maybe_next orelse return -1;
    const rule = maybe_rule orelse return -1;

    next.* = cur.*;
    next.generation = cur.generation +% 1;

    for (0..tile_h) |yy| {
        for (0..tile_w) |xx| {
            const idx = yy * tile_w + xx;
            const self = cur.state[idx];
            const alive = countAlive(&cur.state, @intCast(xx), @intCast(yy), rule.inner_radius, rule.alive_threshold);
            const is_alive = self >= rule.alive_threshold;
            const delta: i32 = if (is_alive)
                if (alive >= rule.survive_lo and alive <= rule.survive_hi) rule.grow_step else -@as(i32, rule.decay_step)
            else if (alive >= rule.birth_lo and alive <= rule.birth_hi)
                rule.grow_step
            else
                -@as(i32, rule.decay_step);
            next.state[idx] = clampU8(@as(i32, self) + delta);
        }
    }
    return 0;
}

pub export fn cm_mnca_tile_encode(
    maybe_t: ?*const Tile,
    maybe_rule: ?*const MncaRule,
    maybe_out: ?[*]u8,
) callconv(.c) usize {
    const t = maybe_t orelse return 0;
    const out = maybe_out orelse return 0;

    @memset(out[0..wire.payload_size], 0);
    wire.writeU16(out[0..2], t.x);
    wire.writeU16(out[2..4], t.y);
    wire.writeU32(out[4..8], t.generation);
    if (maybe_rule) |rule| @memcpy(out[8..12], rule.rule_id[0..]);
    wire.writeU32(out[12..16], tile_cells);
    @memcpy(out[16..][0..tile_cells], t.state[0..]);
    return tile_header_bytes + tile_cells;
}

pub export fn cm_mnca_tile_decode(maybe_payload: ?[*]const u8, payload_len: usize, maybe_out: ?*Tile) callconv(.c) c_int {
    const payload = maybe_payload orelse return -1;
    const out = maybe_out orelse return -1;
    if (payload_len < tile_header_bytes) return -1;
    const state_len = wire.readU32(payload[12..][0..4]);
    if (tile_header_bytes + @as(usize, @intCast(state_len)) > payload_len) return -1;
    if (state_len != tile_cells) return -1;
    out.x = wire.readU16(payload[0..2]);
    out.y = wire.readU16(payload[2..4]);
    out.generation = wire.readU32(payload[4..8]);
    @memcpy(out.state[0..], payload[16..][0..tile_cells]);
    return 0;
}

pub export fn cm_mnca_tile_hash(maybe_t: ?*const Tile, maybe_out_hash: ?[*]u8) callconv(.c) void {
    const t = maybe_t orelse return;
    const out_hash = maybe_out_hash orelse return;
    Sha256.hash(t.state[0..], out_hash[0..32], .{});
}

pub export fn cm_mnca_quorum_init(maybe_q: ?*Quorum) callconv(.c) void {
    const q = maybe_q orelse return;
    q.* = std.mem.zeroes(Quorum);
}

pub export fn cm_mnca_quorum_update(
    maybe_q: ?*Quorum,
    x: u16,
    y: u16,
    generation: u32,
    maybe_tile_hash: ?[*]const u8,
    maybe_sender_mac: ?[*]const u8,
    now_ms: u64,
) callconv(.c) c_int {
    const q = maybe_q orelse return 0;
    const tile_hash = maybe_tile_hash orelse return 0;
    const sender_mac = maybe_sender_mac orelse return 0;

    for (&q.slots) |*s| {
        if (s.valid and (now_ms -% s.first_seen_ms) > quorum_ttl_ms) {
            s.valid = false;
        }
    }

    var slot: ?*QuorumSlot = null;
    var free_idx: ?usize = null;
    for (&q.slots, 0..) |*s, i| {
        if (!s.valid) {
            if (free_idx == null) free_idx = i;
            continue;
        }
        if (s.x == x and s.y == y and s.generation == generation) {
            slot = s;
            break;
        }
    }

    if (slot == null) {
        const idx = free_idx orelse return 0;
        var s = &q.slots[idx];
        s.valid = true;
        s.x = x;
        s.y = y;
        s.generation = generation;
        s.seen_count = 0;
        s.first_seen_ms = now_ms;
        slot = s;
    }

    var s = slot.?;
    for (s.sender_mac[0..s.seen_count]) |mac| {
        if (std.mem.eql(u8, mac[0..], sender_mac[0..6])) return 0;
    }

    const idx = s.seen_count;
    if (idx >= quorum_k + 1) return 0;
    @memcpy(s.tile_hash[idx][0..], tile_hash[0..32]);
    @memcpy(s.sender_mac[idx][0..], sender_mac[0..6]);
    s.seen_count += 1;

    var a: usize = 0;
    while (a < s.seen_count) : (a += 1) {
        var match: u8 = 1;
        var b = a + 1;
        while (b < s.seen_count) : (b += 1) {
            if (std.mem.eql(u8, s.tile_hash[a][0..], s.tile_hash[b][0..])) match += 1;
        }
        if (match >= quorum_k) {
            s.valid = false;
            return 1;
        }
    }
    return 0;
}
